lineapy/plugins/airflow_pipeline_writer.py

```diff
@@ -97,23 +147,40 @@ def task_iris_pred():
         This way, the generated Airflow DAG file opens room for engineers
         to control pipeline runs at a finer level and allows for further customization.
         """
+
         DAG_TEMPLATE = load_plugin_template("airflow_dag_PythonOperator.jinja")
-        if (
-            AirflowDagFlavor[dag_flavor]
-            == AirflowDagFlavor.PythonOperatorPerSession
-        ):
-            task_defs = self.get_session_task_definition()
-        elif (
-            AirflowDagFlavor[dag_flavor]
-            == AirflowDagFlavor.PythonOperatorPerArtifact
-        ):
-            task_defs = self.get_artifact_task_definitions()
-
-        task_params = self.get_task_params_args(task_defs)
+
+        if dag_flavor == AirflowDagFlavor.PythonOperatorPerSession:
+            task_breakdown = DagTaskBreakdown.TaskPerSession
+        elif dag_flavor == AirflowDagFlavor.PythonOperatorPerArtifact:
+            task_breakdown = DagTaskBreakdown.TaskPerArtifact
+
+        # Get task definitions based on dag_flavor
+        task_defs = get_task_definitions(
+            self.artifact_collection,
+            pipeline_name=self.pipeline_name,
+            task_breakdown=task_breakdown,
+        )
+
         task_functions = list(task_defs.keys())
+
+        # Add setup and teardown if local pickle serializer is selected
+        if task_serialization == TaskSerializer.LocalPickle:
+            task_defs["setup"] = get_localpickle_setup_task_definition(
+                self.pipeline_name
+            )
+            task_defs["teardown"] = get_localpickle_teardown_task_definition(
+                self.pipeline_name
+            )
+            # insert in order to task_functions so that setup runs first and teardown runs last
+            task_functions.insert(0, "setup")
+            task_functions.append("teardown")
```

Comment:
In my view, this is actually an example that supports @mingjerli's suggestion to provide the contributor with a graph object instead of a flat/linear dictionary object. That is, if `task_defs` were a graph object, adding these extra tasks (plus their "order" info) would be much more straightforward and syntactically natural.

Also, with `task_defs` being a graph object, the contributor would be spared from manual re-construction of dependency info (see below for an example in our current Airflow pipeline writer) because they can simply extract it from the graph object itself.

```python
# Handle dependencies
dependencies = {
    task_functions[i + 1]: {task_functions[i]}
    for i in range(len(task_functions) - 1)
}
task_graph = TaskGraph(
    nodes=task_functions,
    mapping={tf: tf for tf in task_functions},
    edges=dependencies,
)
task_dependencies = [
    f"{task0} >> {task1}" for task0, task1 in task_graph.graph.edges
]
```

To @mingjerli's point, we already have a custom, simplified graph class (i.e., `TaskGraph`), so we can use it to "package" `TaskDefinition`s as a graph (instead of a dictionary). For this, we do not need to change the current definition of `TaskGraph` because each node in the graph object can have attributes.

Specifically, I am thinking we can do something like:

```python
def get_task_definitions(
    artifact_collection: ArtifactCollection,
    pipeline_name: str,
    task_breakdown: DagTaskBreakdown,
) -> TaskGraph:
    """
    [UPDATED DOCSTRING]
    """
    if task_breakdown == DagTaskBreakdown.TaskAllSessions:
        task_defs = get_allsessions_task_definition(pipeline_name)
    elif task_breakdown == DagTaskBreakdown.TaskPerSession:
        task_defs = get_session_task_definition(artifact_collection, pipeline_name)
    elif task_breakdown == DagTaskBreakdown.TaskPerArtifact:
        task_defs = get_artifact_task_definitions(
            artifact_collection, pipeline_name
        )
    else:
        raise ValueError(
            f"Task breakdown granularity {task_breakdown} is not currently supported."
        )

    # Reconstruct dependency (or get the original one some other way)
    task_functions = list(task_defs.keys())
    dependencies = {
        task_functions[i + 1]: {task_functions[i]}
        for i in range(len(task_functions) - 1)
    }
    task_graph = TaskGraph(
        nodes=task_functions,
        mapping={tf: tf for tf in task_functions},
        edges=dependencies,
    )

    # "Attach" each task definition to the corresponding graph node
    for tf in task_functions:
        task_graph.nodes[tf]["task_definition"] = task_defs[tf]

    return task_graph
```

This way, we can quickly reset the contributor-facing API as a graph (instead of dict); and independently tackle more fundamental refactoring if needed.

Reply:
yeah i totally agree, I'll create a ticket for this